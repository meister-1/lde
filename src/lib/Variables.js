import fs from "fs-sync"

export default class Variables {

    constructor(opts) {

        this.variables = []

        if(opts.file) {

            this.variables = fs.exists(opts.file) ? Variables.createObjList(fs.read(opts.file)) : []
            this.file = opts.file

        } else if (opts.vars) {

            this.variables = opts.vars ? opts.vars : []

        } else if (opts.varsStr) {

            this.variables = opts.varsStr ? Variables.createObjList(opts.varsStr) : []

        }

    }

    static createObjList(variableFile) {

        if(!variableFile) {
            return []
        }

        return variableFile.split("\n").map((varLine) => {
            const varParts = varLine.split("=")
    
            return {key: varParts[0].toUpperCase(), value: varParts[1]}
        }).filter(v => v.key !== undefined && v.key != "")
    
    }

    getObjList() {

        return this.variables
    
    }

    replaceVariables(vars) {

        if(!vars) {
            return this
        }

        this.variables = this.variables.map(({key, value}) => {
            let newValue = value
            vars.forEach(({key, value}) => {
                newValue = newValue.replace(new RegExp(`\\$${key}`, 'g'), value)
            })

            return {key, value: newValue}
        })

        return this

    }

    writeFile(fileName) {

        const varFile = this.variables.map(({key, value}) => `${key}=${value}`).join("\n")

        fs.write(fileName ? fileName : this.file, varFile, function(err) {
            if(err) {
                return console.log(err)
            }
        })
    }
}