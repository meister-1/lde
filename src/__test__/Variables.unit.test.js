import Variables from "../lib/Variables"

jest.mock('fs-sync')

const varsStr = `
FOO=123
ABC=456
PORT=$PORT
`


describe('variables test', () => {

    test('create env vars object list', () => {
      expect(new Variables({varsStr}).getObjList()).toMatchSnapshot()      
    })

    test('write var file', () => {
      new Variables({varsStr}).writeFile(`app-a/.env.tmp`)
      const variables = new Variables({file: "app-a/.env.tmp"})
      expect(new Variables({varsStr}).getObjList()).toEqual(variables.getObjList())
    })

    test('replace vars', () => {
      const expectedResult = new Variables({varsStr}).replaceVariables([
        {key: "PORT", value: "8080"},
        {key: "HOST", value: "meister1.com"}
      ]).getObjList()
      expect(expectedResult).toMatchSnapshot()
    })
})