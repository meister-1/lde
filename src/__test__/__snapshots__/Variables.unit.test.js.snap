// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variables test create env vars object list 1`] = `
Array [
  Object {
    "key": "FOO",
    "value": "123",
  },
  Object {
    "key": "ABC",
    "value": "456",
  },
  Object {
    "key": "PORT",
    "value": "$PORT",
  },
]
`;

exports[`variables test replace vars 1`] = `
Array [
  Object {
    "key": "FOO",
    "value": "123",
  },
  Object {
    "key": "ABC",
    "value": "456",
  },
  Object {
    "key": "PORT",
    "value": "8080",
  },
]
`;
